crates/accelerate/src/commutation_analysis.rs

```diff
@@ -0,0 +1,197 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use pyo3::prelude::PyModule;
+use pyo3::{pyfunction, pymodule, wrap_pyfunction, Bound, PyResult, Python};
+use qiskit_circuit::Qubit;
+
+use crate::commutation_checker::CommutationChecker;
+use hashbrown::HashMap;
+use indexmap::IndexSet;
+use pyo3::prelude::*;
+
+use pyo3::types::{PyDict, PyList};
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+// custom types: IndexSet allows to iterate over the elements in insertion
+// order, but uses the std hasher, which is slower than ahash
+type AIndexSet<T> = IndexSet<T, ::ahash::RandomState>;
+type CommutingNodes = Vec<AIndexSet<NodeIndex>>;
```

Comment:
I'm wondering about the choice of `IndexSet` here is this just for deduplication? The original python code doesn't seem to use a set for anything.

Reply:
@sbrandhsn motivation was to have a fast lookup (we check if the current gate index is already in the commutation set). I can check what the performance is with a normal vector 🙂

Reply:
I was hoping for a runtime speedup coming from a faster execution of line 79 with a set instead of a list :-) I have not benchmarked this though. It might not be a sane optimization as we probably would only get a runtime speedup when we have large sets of commuting gates. It depends on the input circuit but we might assume that large commuting set do not occur in practice. Besides, I thought if the variable name says that the variable is a set, it should be a set instead of a list. :-D I did not consider deduplication.

Reply:
Switched to a `Vec` in: https://github.com/Qiskit/qiskit/pull/12995/commits/70735707d1a1490502d757528399f6b461fbdbda